Wrappers/Python/cil/optimisation/operators/Operator.py

```diff
@@ -118,38 +118,120 @@ def adjoint(self,x, out=None):
+        -------
+
+        dominant eigenvalue: positive:`float`
+        number of iterations: positive:`int`
+            Number of iterations to reach tolerance, if verbose = True
+        eigenvector: DataContainer
+            Corresponding eigenvector of the dominant eigenvalue, if verbose = True
+        list of eigenvalues: :obj:`list`
+            List of eigenvalues
+
+        Examples
+        --------    
+
+        >>> M = np.array([[1.,0],[1.,2.]])
+        >>> Mop = MatrixOperator(M)
+        >>> Mop_norm = Mop.PowerMethod(Mop)
+        >>> Mop_norm
+        2.0000654846240296
+
+        `PowerMethod` is called when we compute the norm of a matrix or a `LinearOperator`.
+
+        >>> Mop_norm = Mop.norm()
+        2.0005647295658866
+        
+
+        Note
+        ----
+
+
+        """
+
+        # Default case: non-symmetric
+        symmetric = False
+
+        # symmetric case  
+        if operator.domain_geometry()==operator.range_geometry():
+            symmetric = True
         
-        # Initialise random
-        if x_init is None:
+        # Initialise random or by the user
+        if initial is None:
             x0 = operator.domain_geometry().allocate('random')
         else:
-            x0 = x_init.copy()
+            x0 = initial.copy()
+
+        # Normalize first eigenvector
+        x0_norm = x0.norm()
+        x0 /= x0_norm
             
-        x1 = operator.domain_geometry().allocate()
+        # allocate space on the range geometry
         y_tmp = operator.range_geometry().allocate()
-        s = numpy.zeros(iterations)
-        # Loop
-        for it in numpy.arange(iterations):
-            operator.direct(x0,out=y_tmp)
-            operator.adjoint(y_tmp,out=x1)
-            x1norm = x1.norm()
-            if hasattr(x0, 'squared_norm'):
-                s[it] = x1.dot(x0) / x0.squared_norm()
+
+        # initial guess for dominant eigenvalue
+        eig_old = 1.
+
+        # list of eigenvalues
+        eig_list = []
+
+        for i in range(iterations):
+
+            if symmetric:                
+                operator.direct(x0, y_tmp)
+                # in-place?
+                x0 = y_tmp.copy()
+            else:
+                operator.direct(x0, out = y_tmp)
+                operator.adjoint(y_tmp,out=x0)
+            
+            # Get eigenvalue using Rayleigh quotient: denominator=1, due to normalization
+            x0_norm = x0.norm()
+            x0 /= x0_norm
+
+            if symmetric:
+                eig_new =  numpy.abs(x0_norm)
             else:
-                x0norm = x0.norm()
-                s[it] = x1.dot(x0) / (x0norm * x0norm) 
-            x1.multiply((1.0/x1norm), out=x0)
-        return numpy.sqrt(s[-1]), numpy.sqrt(s), x0
+                eig_new = numpy.sqrt(numpy.abs(x0_norm))
+            
+            # Stopping criterion of two consecutive eigenvalues
+            if i>0 and numpy.abs(eig_new - eig_old) < tolerance :
+                break

```

Comment:
It will check line 224 at least one time so `eig_new` will be appended in `eig_list`.

Reply:
Yes, the one before it hits the stopping criteria will. But I don't see any harm in including every value that's been calculated. The `append` line just has to move above the stopping criteria for this.

I'd also consider making it a `while` loop rather than using a break.

Reply:
I move `append ` above the stopping criterion. The `while` would also require a `break` since we need to check two conditions and I prefer not doing it at the beginning since I have not computed `eig_new`.

Reply:
I still feel that would be clearer to read/maintain as you don't have to look at the for loop set-up and the break statement to see the stopping criteria. I feel this would be clearer:



```python
diff = np.finfo('d').max
i = 0
while(i < iteration and diff > tolerance ):
    ....
    diff = numpy.abs(eig_new - eig_old)
    i+=1
```

Reply:
Changed to a `while` loop